src/script/interpreter.cpp

```diff
@@ -1416,9 +1416,26 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
 template class GenericTransactionSignatureChecker<CTransaction>;
 template class GenericTransactionSignatureChecker<CMutableTransaction>;
 
+static bool ExecuteWitnessProgram(std::vector<std::vector<unsigned char>> stack, const CScript& scriptPubKey, unsigned int flags, SigVersion sigversion, const BaseSignatureChecker& checker, ScriptError* serror)
```

Comment:
nit: Maybe `stack` passing could be clarified by passing as const reference and instantiating the local within? Or a doc commenting that it's an in var.

Reply:
Passing it as const reference would require an unnecessary copy. I'll add a comment.

Reply:
Sounds good re comment. Doesn't the copy occur in either case? AFAIK pass by value requires a copy to have a new instance available for modification.

Reply:
Pass by value is a terrible name.

It's just a variable on the callee side that is constructed in place by the caller. The call sites construct temporaries, so the move constructor will be invoked to construct the parameter.

Reply:
Everything else is called `Eval` rather than `Execute`

Reply:
I didn't want to call it ...Eval, as it's not just evaluating a script (EvalScript just executes opcodes, and modifies a stack; this function is actually closer to VerifyScript which maps it to a true/false success overall).

Reply:
Concerning the `stack` parameter: would it make sense to declare it as an rvalue parameter (&&) and pass constructed temporaries in both calls (would need to change `witness.stack` to `{witness.stack.begin(), witness.stack.end()}`, to ensure and clarify that the move-constructor is _always_ involved? Just an idea following the "explicit is better than implicit" principle.

Reply:
@theStack Ok, better like this?

Reply:
That was more of a suggestion/question than a concrete wish (you probably have way more C++ experience), but personally I'd prefer it, yes. Interestingly enough, on the first call explicitely calling the constructor would not be needed (passing `{witness.stack.begin(), witness.stack.end() - 1}` is sufficient), while on the second call only `{witness.stack}` would lead to an error and an explicit constructor call (`std::vector<std::vector<unsigned char>>(...)`) is needed.

Reply:
Yes, I'm aware it wasn't necessary to invoke the constructor explicitly, but it's probably clearer this way (semantically it's equivalent).

Reply:
You could do:

```diff
-static bool ExecuteWitnessProgram(std::vector<std::vector<unsigned char>>&& stack, const CScript& scriptPubKey, unsigned int flags, SigVersion sigversion, const BaseSignatureChecker& checker, ScriptError* serror)
+static bool ExecuteWitnessProgram(std::vector<valtype>::const_iterator begin, std::vector<valtype>::const_iterator end, const CScript& scriptPubKey, unsigned int flags, SigVersion sigversion, const BaseSignatureChecker& checker, ScriptError* serror)
 {
+    std::vector<valtype> stack{begin, end};
```

and just pass the iterators in directly.

Reply:
Done, as that makes it also more compatible with #13062.